data_rentgen/server/utils/lineage_response.py

```diff
@@ -232,3 +230,122 @@ def _get_indirect_column_lineage(column_lineage_by_source_target_id: dict[tuple,
             column_lineage_relation.fields = fields
             relations.append(column_lineage_relation)
     return sorted(relations, key=lambda x: (str(x.from_.id), str(x.to.id)))
+
+
+def _get_datasets(
+    raw_datasets: dict[int, Dataset],
+    outputs: dict[Any, OutputRow],
+    inputs: dict[Any, InputRow],
+) -> dict[str, DatasetResponseV1]:
+    datasets: dict[str, DatasetResponseV1] = defaultdict()
+    for dataset in raw_datasets.values():
+        schema: DatasetSchemaV1 | None = None
+        schema = _get_latest_io_schema(dataset, outputs)
+        if not schema:
+            schema = _get_latest_io_schema(dataset, inputs)
+        datasets[str(dataset.id)] = DatasetResponseV1(
+            id=str(dataset.id),
+            location=LocationResponseV1.model_validate(dataset.location),
+            name=dataset.name,
+            format=dataset.format,
+            schema=schema,
+        )
+    return datasets
+
+
+def _get_latest_io_schema(dataset, relations: Mapping[Any, OutputRow | InputRow]) -> DatasetSchemaV1 | None:
+    oldest_relation, newest_relation = None, None
+    dataset_schema = None
+    for relation in relations.values():
+        if dataset.id == relation.dataset_id:
+            if (oldest_relation is None or relation.created_at <= oldest_relation.created_at) and relation.schema_id:
+                if oldest_relation and relation.created_at == oldest_relation.created_at:
+                    oldest_relation = max(oldest_relation, relation, key=lambda x: x.schema_id)  # type: ignore[assignment]
+                else:
+                    oldest_relation = relation
+            if (newest_relation is None or relation.created_at >= newest_relation.created_at) and relation.schema_id:
+                if newest_relation and relation.created_at == newest_relation.created_at:
+                    newest_relation = max(newest_relation, relation, key=lambda x: x.schema_id)  # type: ignore[assignment]
+                else:
+                    newest_relation = relation
+    if oldest_relation and newest_relation is None:
+        dataset_schema = DatasetSchemaV1.model_validate(oldest_relation.schema)
+    elif oldest_relation is None and newest_relation:
+        dataset_schema = DatasetSchemaV1.model_validate(newest_relation.schema)
+    elif oldest_relation and newest_relation:
+        dataset_schema = DatasetSchemaV1.model_validate(newest_relation.schema)
+        dataset_schema.relevance_type = (
+            "EXACT_MATCH" if oldest_relation.schema_id == newest_relation.schema_id else "LATEST_KNOWN"
+        )
+    else:
+        return None
+
+    return dataset_schema
+
+
+def _get_datasets_with_dataset_granularity(
```

Comment:
This could be simplified as well, by first getting latest schema for each dataset, and then by converting DB models to response schemas